### ISSUE TITLE
not authorized to access endpoint

### ISSUE THREAD

**Issue by tisoft** [LukasK32/fleeting-plugin-proxmox#6]
The proxmox fleeting plugin works fine for a while and then suddenly, I get `not authorized to access endpoint` errors.

This might be some misconfiguration by me, but I couldn't find any error. Is there any session timeout, that could be at work here? Or could you give me some hints in where to look for any clues?

After restarting the gitlab runner, the fleeting plugin works again (until it stops working again). 

The following log shows, that it s working until 16:19:39  and fails starting at 18:07:48:

```
Oct 07 16:19:06 proxmoxrunnertest gitlab-runner[1681]: Job succeeded                                       duration_s=253.15364596 job=317523 project=44 runner=pLXizZ87_
Oct 07 16:19:06 proxmoxrunnertest gitlab-runner[1681]: Appending trace to coordinator...ok                 code=202 job=317523 job-log=0-4091 job-status=running runner=pLXizZ87_ sent-log=4015-4090 status=202 Accepted update-interval=3s
Oct 07 16:19:06 proxmoxrunnertest gitlab-runner[1681]: Updating job...                                     bytesize=4091 checksum=crc32:5e83da21 job=317523 runner=pLXizZ87_
Oct 07 16:19:06 proxmoxrunnertest gitlab-runner[1681]: Submitting job to coordinator...ok                  bytesize=4091 checksum=crc32:5e83da21 code=200 job=317523 job-status=success runner=pLXizZ87_ update-interval=0s
Oct 07 16:19:06 proxmoxrunnertest gitlab-runner[1681]: Removed job from processing list                    builds=0 job=317523 max_builds=1 project=44 repo_url=https://XXX/playground/linux-docker-build-test.git time_in_queue_seconds=0
Oct 07 16:19:06 proxmoxrunnertest gitlab-runner[1681]: instance marked for removal                         instance=108 reason=instance reached max use count runner=pLXizZ87_ subsystem=taskscaler
Oct 07 16:19:07 proxmoxrunnertest gitlab-runner[1681]: decreasing instances                                amount=1 group=gitlab runner=pLXizZ87_ subsystem=taskscaler
Oct 07 16:19:07 proxmoxrunnertest gitlab-runner[1681]: removing instance                                   runner=pLXizZ87_ subsystem=fleeting-plugin.plugin timestamp=2024-10-07T16:19:07.102+0200 vmid=108
Oct 07 16:19:17 proxmoxrunnertest gitlab-runner[1681]: instance update                                     group=gitlab id=108 runner=pLXizZ87_ state=deleting subsystem=taskscaler
Oct 07 16:19:27 proxmoxrunnertest gitlab-runner[1681]: collector found instance to remove                  name=fleeting-removing runner=pLXizZ87_ subsystem=fleeting-plugin.plugin timestamp=2024-10-07T16:19:27.239+0200 vmid=108
Oct 07 16:19:39 proxmoxrunnertest gitlab-runner[1681]: instance pruned                                     group=gitlab id=108 lifetime=7m35.36057442s runner=pLXizZ87_ subsystem=taskscaler
Oct 07 18:07:48 proxmoxrunnertest gitlab-runner[1681]: ERROR: reconcile                                    err=rpc error: code = Unknown desc = failed to get pool id='gitlab': not authorized to access endpoint runner=pLXizZ87_ subsystem=taskscaler
Oct 07 18:07:52 proxmoxrunnertest gitlab-runner[1681]: ERROR: collector failed to list instances           err=failed to get pool id='gitlab': not authorized to access endpoint runner=pLXizZ87_ subsystem=fleeting-plugin.plugin timestamp=2024-10-07T18:07:52.305+0200
Oct 07 18:08:48 proxmoxrunnertest gitlab-runner[1681]: ERROR: reconcile                                    err=rpc error: code = Unknown desc = failed to get pool id='gitlab': not authorized to access endpoint runner=pLXizZ87_ subsystem=taskscaler
Oct 07 18:08:55 proxmoxrunnertest gitlab-runner[1681]: ERROR: collector failed to list instances           err=failed to get pool id='gitlab': not authorized to access endpoint runner=pLXizZ87_ subsystem=fleeting-plugin.plugin timestamp=2024-10-07T18:08:55.330+0200
Oct 07 18:09:48 proxmoxrunnertest gitlab-runner[1681]: ERROR: reconcile                                    err=rpc error: code = Unknown desc = failed to get pool id='gitlab': not authorized to access endpoint runner=pLXizZ87_ subsystem=taskscaler
Oct 07 18:09:58 proxmoxrunnertest gitlab-runner[1681]: ERROR: collector failed to list instances           err=failed to get pool id='gitlab': not authorized to access endpoint runner=pLXizZ87_ subsystem=fleeting-plugin.plugin timestamp=2024-10-07T18:09:58.352+0200
Oct 07 18:10:48 proxmoxrunnertest gitlab-runner[1681]: ERROR: reconcile                                    err=rpc error: code = Unknown desc = failed to get pool id='gitlab': not authorized to access endpoint runner=pLXizZ87_ subsystem=taskscaler
Oct 07 18:11:01 proxmoxrunnertest gitlab-runner[1681]: ERROR: collector failed to list instances           err=failed to get pool id='gitlab': not authorized to access endpoint runner=pLXizZ87_ subsystem=fleeting-plugin.plugin timestamp=2024-10-07T18:11:01.373+0200
```

I can see the gitlab user in the proxmox log at the time the last successful actions were taken.

```
Oct 07 16:19:27 pve pvedaemon[1505]: <gitlab@pve> end task UPID:pve:0000187E:00015237:6703EDEF:qmstop:108:gitlab@pve: OK
Oct 07 16:19:28 pve systemd[1]: 108.scope: Deactivated successfully.
Oct 07 16:19:28 pve systemd[1]: 108.scope: Consumed 3min 39.770s CPU time.
Oct 07 16:19:28 pve pvedaemon[1503]: <root@pam> end task UPID:pve:00001847:00014D16:6703EDE2:vncproxy:108:root@pam: OK
Oct 07 16:19:28 pve qmeventd[6300]: Starting cleanup for 108
Oct 07 16:19:28 pve qmeventd[6300]: Finished cleanup for 108
Oct 07 16:19:37 pve pvedaemon[1505]: <gitlab@pve> starting task UPID:pve:000018DE:00015628:6703EDF9:qmdestroy:108:gitlab@pve:
Oct 07 16:19:37 pve pvedaemon[6366]: destroy VM 108: UPID:pve:000018DE:00015628:6703EDF9:qmdestroy:108:gitlab@pve:
Oct 07 16:19:38 pve pvedaemon[1505]: <gitlab@pve> end task UPID:pve:000018DE:00015628:6703EDF9:qmdestroy:108:gitlab@pve: OK
```

After that I don't see any mention. Especially not at the time, the errors occur on the runner:
```
Oct 07 17:41:45 pve pmxcfs[1328]: [status] notice: received log
Oct 07 17:44:21 pve pveproxy[38732]: worker exit
Oct 07 17:44:21 pve pveproxy[1531]: worker 38732 finished
Oct 07 17:44:21 pve pveproxy[1531]: starting 1 worker(s)
Oct 07 17:44:21 pve pveproxy[1531]: worker 47315 started
Oct 07 17:51:23 pve pvedaemon[39994]: <root@pam> successful auth for user 'root@pam'
Oct 07 17:56:46 pve pmxcfs[1328]: [status] notice: received log
Oct 07 18:01:09 pve pveproxy[41343]: worker exit
Oct 07 18:01:10 pve pveproxy[1531]: worker 41343 finished
Oct 07 18:01:10 pve pveproxy[1531]: starting 1 worker(s)
Oct 07 18:01:10 pve pveproxy[1531]: worker 56061 started
Oct 07 18:05:36 pve pmxcfs[1328]: [dcdb] notice: data verification successful
Oct 07 18:06:23 pve pvedaemon[45032]: <root@pam> successful auth for user 'root@pam'
Oct 07 18:08:14 pve pvedaemon[39994]: worker exit
Oct 07 18:08:14 pve pvedaemon[1502]: worker 39994 finished
Oct 07 18:08:14 pve pvedaemon[1502]: starting 1 worker(s)
Oct 07 18:08:14 pve pvedaemon[1502]: worker 59616 started
Oct 07 18:11:49 pve pmxcfs[1328]: [status] notice: received log
Oct 07 18:17:01 pve CRON[64238]: pam_unix(cron:session): session opened for user root(uid=0) by (uid=0)
Oct 07 18:17:01 pve CRON[64239]: (root) CMD (cd / && run-parts --report /etc/cron.hourly)
Oct 07 18:17:01 pve CRON[64238]: pam_unix(cron:session): session closed for user root
```

Thanks for the great project! I hope you can help me here. 😄

**Comment by tisoft** [comment]
This seems to occur every 2 hours:

```
Oct 07 20:22:24 proxmoxrunnertest gitlab-runner[1891]: instance discovery                                  cause=requested group=gitlab id=103 runner=pLXizZ87_ state=creating subsystem=taskscaler
Oct 07 20:22:27 proxmoxrunnertest gitlab-runner[1891]: instance update                                     group=gitlab id=103 runner=pLXizZ87_ state=running subsystem=taskscaler
Oct 07 20:22:27 proxmoxrunnertest gitlab-runner[1891]: instance is ready                                   instance=103 runner=pLXizZ87_ subsystem=taskscaler took=228.160733ms
Oct 07 20:22:36 proxmoxrunnertest gitlab-runner[1891]: instance pruned                                     group=gitlab id=102 lifetime=2m23.919441013s runner=pLXizZ87_ subsystem=taskscaler
Oct 07 22:20:45 proxmoxrunnertest gitlab-runner[1891]: ERROR: reconcile                                    err=rpc error: code = Unknown desc = failed to get pool id='gitlab': not authorized to access endpoint runner=pLXizZ87_ subsystem=taskscaler
Oct 07 22:20:49 proxmoxrunnertest gitlab-runner[1891]: ERROR: collector failed to list instances           err=failed to get pool id='gitlab': not authorized to access endpoint runner=pLXizZ87_ subsystem=fleeting-plugin.plugin timestamp=2024-10-07T22:20:49.704+0200
Oct 07 22:21:45 proxmoxrunnertest gitlab-runner[1891]: ERROR: reconcile                                    err=rpc error: code = Unknown desc = failed to get pool id='gitlab': not authorized to access endpoint runner=pLXizZ87_ subsystem=taskscaler
```

```
Oct 08 07:41:48 proxmoxrunnertest gitlab-runner[2183]: instance discovery                                  cause=requested group=gitlab id=102 runner=pLXizZ87_ state=creating subsystem=taskscaler
Oct 08 07:41:51 proxmoxrunnertest gitlab-runner[2183]: instance update                                     group=gitlab id=102 runner=pLXizZ87_ state=running subsystem=taskscaler
Oct 08 07:41:51 proxmoxrunnertest gitlab-runner[2183]: instance is ready                                   instance=102 runner=pLXizZ87_ subsystem=taskscaler took=192.458232ms
Oct 08 07:42:00 proxmoxrunnertest gitlab-runner[2183]: instance pruned                                     group=gitlab id=103 lifetime=2m34.425168173s runner=pLXizZ87_ subsystem=taskscaler
Oct 08 09:40:09 proxmoxrunnertest gitlab-runner[2183]: ERROR: reconcile                                    err=rpc error: code = Unknown desc = failed to get pool id='gitlab': not authorized to access endpoint runner=pLXizZ87_ subsystem=taskscaler
Oct 08 09:40:13 proxmoxrunnertest gitlab-runner[2183]: ERROR: collector failed to list instances           err=failed to get pool id='gitlab': not authorized to access endpoint runner=pLXizZ87_ subsystem=fleeting-plugin.plugin timestamp=2024-10-08T09:40:13.800+0200
Oct 08 09:41:09 proxmoxrunnertest gitlab-runner[2183]: ERROR: reconcile                                    err=rpc error: code = Unknown desc = failed to get pool id='gitlab': not authorized to access endpoint runner=pLXizZ87_ subsystem=taskscaler
```

This seems to be hardcoded in proxmox:

https://git.proxmox.com/?p=pve-access-control.git;a=blob;f=src/PVE/AccessControl.pm;h=13065764ec20a8edcc4678f5feede3ecd7daebf8;hb=HEAD#l57

Maybe using API tokens instead of a password would work here?

**Comment by LukasK32** [comment]
* API tokens should work fine but there is no way to use them with this plugin at the moment - added to my TODOs ;)
* I will add a patch to refresh the session periodically to fix this

**Comment by LukasK32** [comment]
* Prepared and tested the patch, released as [`v1.1.1`](https://github.com/LukasK32/fleeting-plugin-proxmox/releases/tag/v1.1.1)
* Let me know if it works for you :)

**Comment by tisoft** [comment]
Has now worked flawlessly for the last 2 weeks!

Thanks for the fix!
